Obj damaging system, acid damage, and fire damage refactor

phil235 opened the pull request:
Please refer to #20867 and #20870 for a easier view of the changes. Those two PRs show all meaningful changes (hopefully) and doesn't show the files changed with just 3 lines changed.

This PR does three things:
- It makes all children of /obj/ use the same damage system.
  Previously to make your new machine/structure be destroyable you needed to give it a var/health, and its own version of many damage related proc such as bullet_act(), take_damage(), attacked_by(), attack_animal(), attack_hulk(), ex_act(), etc... But now, all /obj/ use the same version of those procs at the /obj/ level in code/game/obj_defense.dm. All these obj share the same necessary vars: obj_integrity (health), max_integrity, integrity_failure (optional, below that health level failure happens), and the armor list var which was previously only for items, as well as the resistance_flags bitfield. When you want your new object to be destroyable, you only have to give it a value for those vars and maybe override one proc if you want a special behavior but that's it. This reorganization removes a lot of copypasta (most bullet_act() version for each obj were nearly identical). Two new elements are added to the armor list var: fire and acid armor values.
  How much damage an obj take depends on the armor value for each damage category. But some objects are INDESTRUCTIBLE and simply never take any damage no matter the type.
  The armor categories are: 
  -melee(punches, item attacks, xeno/animal/hulk attacks, blob attacks, thrown weapons)
  -bullet
  -laser
  -energy (used by projectiles like ionrifle, taser, and also by EMPs)
  -bio (unused for this, only here because clothes use them when worn)
  -rad (same)
  -bomb (self-explanatory)
  -fire (for fire damage, not for heat damage though)
  -acid
  For machines and structures, when their health reaches zero the object is not just deleted but gets somewhat forcedeconstructed (the proc used is shared with the actual deconstruction system) which can drops things. To not frustrates players most of these objects drop most of the elements necessary to rebuild them (think window dropping shards). Machines drop a machine frame and all components for example (but the frame can then be itself smashed to pieces).
  For clothes, when they are damaged, they get a "damaged" overlay, which can also be seen when worn, similar to the "bloody" overlay.
- It refactors acid. See #20537.
  Some objects are ACID_PROOF and take no damage from acid, while others take varying amounts
  of damage depending on their acid armor value. Some objects are even UNACIDABLE, no acid effect can even land on them. Acid on objects can be washed off using water.
- It changes some aspect of damage from fires.
  All /obj/ can now take fire damage and be flammable, instead of just items. And instead of having just FLAMMABLE objs that become ON_FIRE as soon as some fire touch them (paper), we now have objects that are non flammable but do take damage from fire and become ashes if their health reaches zero (only for items). The damage taken varies depending on the obj's fire armor value and total health. There's also still obj and items that are FIRE_PROOF (although some might still be melted by lava if they're not LAVA_PROOF).
  When a mob is on fire, its clothes now take fire damage and can turn to ashes. Similarly, when a mob takes melee damages, its clothes gets damaged a bit and can turn to shreds. You can repair clothes with cloth that is produceable by botany's biogenerator.
- It also does many minor things:
  - Clicking a structure/machine with an item on help intent never results in an attack (so you don't destroy a structure while trying to figure out which tool to use).
  - I moved a lot of objects away from /obj/effect, it should only be used for visual effects, decals and stuff, not for things you can hit and destroy.
  - I tweaked a bit how clothes shredding from bombs work.
  - I made a machine or structure un/anchorable with the wrench, I don't remember which object...
  - Since I changed the meaning of the FIRE_PROOF bitflag to actually mean fire immune, I'm buffing the slime extract that you apply on items to make them fire proof. well now they're really 100% fire proof!
  - animals with environment_smash = 1 no longer one-hit destroy tables and stuff, we give them a decent obj_damage value so they can destroy most obj relatively fast depending on the animal.
  - Probably a million things I forgot.

If you want to know how the damage system works all you need is the three obj vars "obj_integrity", "max_integrity", "integrity_failure", as well as the armor list var and the resistance_flags bitfield, and read the file obj_defense.dm

:cl: phil235
add: A whole bunch of structures and machines can now be broken and destroyed using conventional attack methods.
add: The effect of fire on objects is overhauled. When on fire, your external clothes now take fire damage and can become ashes, unless fire proof. All objects caught in a fire can now take fire damage, instead of just flammable ones, unless fire proof.
add: Acid effect is overhauled. Splashing acid on something puts a green effect on it and start slowly melting it. Throwing acid on mobs puts acid on their clothes. Acid can destroy structures and items alike, as long as they're not acid proof. The more acid put on an object, the faster it melts. Walking on an acid puddle puts acid on your shoes or burns your bare feet. Picking up an item with acid without adequate glove protection burns your hand. Acid can be washed away with any sort of water source (extinguisher, shower, water glass,etc...).
add: When a mob receives melee attacks, its clothes also get damaged. Clothes with enough damage get a visual effect. Damaged clothes can be repaired with cloth (produced with botany's biogenerator). Clothes that receives too much melee damage become shreds.
tweak: Clicking a structure/machine with an item on help intent can never result in an attack.
tweak: Hostile animals that are environment destroyers no longer destroys tables&closets in one hit. It now takes several hits depending on the animal.
/:cl:

phil235 committed:
Refactors how acid works. Alien corrosive acid and reagent acids now use the same system. Acid applied to an obj or turf now does not act instantly, it is processed so its target slowly melts away, with an acid sprite on it.
Only alien acid can melt turfs, for balance purposes.
The acid_state var for objs distinguishes the different behavior with acid:
UNACIDABLE: the acid cannot even stick to the obj, let alone damage it, no acid effect appears on the obj.
ACID_PROOF the obj is immune to acid, but if it receive acid it will show acid icon until the acid stops processing.
ACIDABLE: All objs that can be melted. How long an object takes to completely melt is given by the acid_resistance var.

Running on an acid puddle transfer some acid to your shoes or hurt your legs.
Objects that sit on an acid puddle get some acid transfered to them (if puddle has enough acid).

Splashing more acid on an obj or turf that's already being acided transfer more acid to them, but the max total acid volume in an acid puddle is limited to 300u (so stronger acids keep an advantage over weaker acids).

phil235 committed:
Make the damage system for children of /obj centralized: structures, items, machines all use the same procs and vars to handle damage.
I changed a few obj/effect/ objects into obj/structure since it made more sense (blob, cocoon, vines)
I moved fire and acid processing into their own subsystem.
Some other things still in progress.

phil235 committed:
ninja energy net into a structure.

phil235 committed:
Lots of work on children of obj to make them damageable, or adapt their destruction code to the new damage system.

phil235 committed:
A lot of work related to damaging objs. Adjusting values, tweaking what some obj drop when destroyed.
Airlock can now be damaged, broken, and destroyed.
Some work on disposal and gas pipes, as well as cable (so they aren't harmed by acid smoke or fire when hidden under a floor).

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into damagerino

# Conflicts:
#	code/__DEFINES/misc.dm
#	code/_onclick/hud/screen_objects.dm
#	code/controllers/subsystem/processing/objects.dm
#	code/game/gamemodes/blob/blob_report.dm
#	code/game/gamemodes/blob/blobs/core.dm
#	code/game/gamemodes/blob/theblob.dm
#	code/game/gamemodes/changeling/powers/mutations.dm
#	code/game/gamemodes/clock_cult/clock_items.dm
#	code/game/gamemodes/clock_cult/clock_ratvar.dm
#	code/game/gamemodes/clock_cult/clock_structures.dm
#	code/game/gamemodes/cult/cult_structures.dm
#	code/game/gamemodes/cult/runes.dm
#	code/game/gamemodes/gang/dominator.dm
#	code/game/gamemodes/handofgod/structures.dm
#	code/game/gamemodes/miniantags/bot_swarm/swarmer.dm
#	code/game/gamemodes/wizard/artefact.dm
#	code/game/machinery/camera/camera.dm
#	code/game/machinery/computer/computer.dm
#	code/game/machinery/deployable.dm
#	code/game/machinery/doors/door.dm
#	code/game/machinery/doors/windowdoor.dm
#	code/game/machinery/machinery.dm
#	code/game/machinery/newscaster.dm
#	code/game/machinery/porta_turret/portable_turret.dm
#	code/game/machinery/shieldgen.dm
#	code/game/machinery/vending.dm
#	code/game/mecha/combat/combat.dm
#	code/game/mecha/combat/durand.dm
#	code/game/mecha/combat/gygax.dm
#	code/game/mecha/combat/marauder.dm
#	code/game/mecha/combat/phazon.dm
#	code/game/mecha/combat/reticence.dm
#	code/game/mecha/mecha.dm
#	code/game/mecha/mecha_defense.dm
#	code/game/mecha/working/ripley.dm
#	code/game/objects/effects/alien_acid.dm
#	code/game/objects/effects/anomalies.dm
#	code/game/objects/effects/decals/Cleanable/aliens.dm
#	code/game/objects/effects/decals/Cleanable/humans.dm
#	code/game/objects/effects/decals/Cleanable/misc.dm
#	code/game/objects/effects/decals/Cleanable/robots.dm
#	code/game/objects/effects/decals/cleanable.dm
#	code/game/objects/effects/effect_system/effects_foam.dm
#	code/game/objects/effects/effects.dm
#	code/game/objects/effects/glowshroom.dm
#	code/game/objects/effects/overlays.dm
#	code/game/objects/effects/spiders.dm
#	code/game/objects/items.dm
#	code/game/objects/items/devices/instruments.dm
#	code/game/objects/items/robot/robot_upgrades.dm
#	code/game/objects/items/stacks/medical.dm
#	code/game/objects/items/stacks/sheets/mineral.dm
#	code/game/objects/items/toys.dm
#	code/game/objects/items/weapons/chrono_eraser.dm
#	code/game/objects/items/weapons/grenades/grenade.dm
#	code/game/objects/items/weapons/holosign_creator.dm
#	code/game/objects/items/weapons/paint.dm
#	code/game/objects/items/weapons/storage/backpack.dm
#	code/game/objects/items/weapons/storage/briefcase.dm
#	code/game/objects/objs.dm
#	code/game/objects/structures/aliens.dm
#	code/game/objects/structures/artstuff.dm
#	code/game/objects/structures/beds_chairs/bed.dm
#	code/game/objects/structures/beds_chairs/chair.dm
#	code/game/objects/structures/bedsheet_bin.dm
#	code/game/objects/structures/crates_lockers/closets/cardboardbox.dm
#	code/game/objects/structures/crates_lockers/closets/gimmick.dm
#	code/game/objects/structures/crates_lockers/closets/job_closets.dm
#	code/game/objects/structures/crates_lockers/closets/secure/bar.dm
#	code/game/objects/structures/crates_lockers/closets/secure/personal.dm
#	code/game/objects/structures/crates_lockers/closets/secure/security.dm
#	code/game/objects/structures/displaycase.dm
#	code/game/objects/structures/fireaxe.dm
#	code/game/objects/structures/flora.dm
#	code/game/objects/structures/girders.dm
#	code/game/objects/structures/grille.dm
#	code/game/objects/structures/lattice.dm
#	code/game/objects/structures/mineral_doors.dm
#	code/game/objects/structures/signs.dm
#	code/game/objects/structures/tables_racks.dm
#	code/game/objects/structures/window.dm
#	code/game/turfs/simulated/floor/plating.dm
#	code/modules/atmospherics/environmental/LINDA_fire.dm
#	code/modules/atmospherics/machinery/portable/canister.dm
#	code/modules/awaymissions/gateway.dm
#	code/modules/clothing/clothing.dm
#	code/modules/clothing/glasses/glasses.dm
#	code/modules/clothing/spacesuits/hardsuit.dm
#	code/modules/clothing/suits/armor.dm
#	code/modules/clothing/suits/cloaks.dm
#	code/modules/clothing/under/ties.dm
#	code/modules/events/spacevine.dm
#	code/modules/events/wizard/greentext.dm
#	code/modules/food_and_drinks/drinks/drinks/drinkingglass.dm
#	code/modules/holodeck/area_copy.dm
#	code/modules/library/lib_codex_gigas.dm
#	code/modules/library/lib_items.dm
#	code/modules/mining/lavaland/ash_flora.dm
#	code/modules/mining/lavaland/necropolis_chests.dm
#	code/modules/mining/mine_items.dm
#	code/modules/mining/money_bag.dm
#	code/modules/mob/inventory.dm
#	code/modules/mob/living/bloodcrawl.dm
#	code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm
#	code/modules/mob/living/carbon/human/human_defense.dm
#	code/modules/mob/living/carbon/monkey/monkey_defense.dm
#	code/modules/mob/living/simple_animal/friendly/farm_animals.dm
#	code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm
#	code/modules/modular_computers/file_system/programs/antagonist/revelation.dm
#	code/modules/ninja/energy_katana.dm
#	code/modules/ninja/suit/gloves.dm
#	code/modules/ninja/suit/head.dm
#	code/modules/ninja/suit/mask.dm
#	code/modules/ninja/suit/n_suit_verbs/energy_net_nets.dm
#	code/modules/ninja/suit/shoes.dm
#	code/modules/ninja/suit/suit.dm
#	code/modules/paperwork/contract.dm
#	code/modules/paperwork/paper.dm
#	code/modules/paperwork/paper_cutter.dm
#	code/modules/paperwork/paperplane.dm
#	code/modules/paperwork/photocopier.dm
#	code/modules/paperwork/photography.dm
#	code/modules/power/apc.dm
#	code/modules/power/gravitygenerator.dm
#	code/modules/power/lighting.dm
#	code/modules/power/singularity/containment_field.dm
#	code/modules/power/singularity/singularity.dm
#	code/modules/power/solar.dm
#	code/modules/power/supermatter/supermatter.dm
#	code/modules/projectiles/projectile.dm
#	code/modules/projectiles/projectile/energy.dm
#	code/modules/recycling/disposal-structures.dm
#	code/modules/research/xenobiology/xenobiology.dm
#	code/modules/ruins/objects_and_mobs/necropolis_gate.dm
#	code/modules/shuttle/shuttle.dm
#	code/modules/shuttle/special.dm
#	code/modules/spells/spell_types/ethereal_jaunt.dm
#	code/modules/surgery/bodyparts/dismemberment.dm
#	code/modules/surgery/organs/augments_arms.dm
#	icons/effects/effects.dmi
#	icons/obj/chemical.dmi

phil235 committed:
Some tweaks and fixing some merge conflicts errors.

RemieRichards commented on code/game/mecha/working/ripley.dm:
```diff
@@ -29,7 +28,7 @@
 	for(var/i=1, i <= hides, i++)
 		new /obj/item/stack/sheet/animalhide/goliath_hide(loc) //If a goliath-plated ripley gets killed, all the plates drop
-	damage_absorption["brute"] =  initial(damage_absorption["brute"])
+	armor["melee"] =  initial(armor["melee"]) //phil235 changed brute to melee!
```
fun fact, this line does nothing.

RemieRichards commented on code/game/objects/obj_defense.dm:
```diff
@@ -0,0 +1,257 @@
+
+/obj/proc/empty_object_contents(burn = 0, new_loc = src.loc)
+	for(var/obj/item/Item in contents) //Empty out the contents
+		Item.loc = new_loc
```
all ```thing.loc = X``` should be ```thing.forceMove(X)```

RemieRichards commented on code/game/objects/structures/lattice.dm:
```diff
@@ -30,37 +32,24 @@
 
-/obj/structure/lattice/Deconstruct()
+/obj/structure/lattice/deconstruct()
 	stored.loc = get_turf(src)
```
all ```thing.loc = X``` should be ```thing.forceMove(X)```

RemieRichards commented on code/modules/hydroponics/hydroponics.dm:
```diff
@@ -773,7 +774,7 @@
-			health = myseed.endurance
+			plant_health = myseed.endurance
 			lastcycle = world.time
 			O.loc = src
```
all ```thing.loc = X``` should be ```thing.forceMove(X)```

RemieRichards commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -219,7 +219,7 @@
 				L.embedded_objects -= I
-				L.take_damage(I.embedded_unsafe_removal_pain_multiplier*I.w_class)//It hurts to rip it out, get surgery you dingus.
+				L.receive_damage(I.embedded_unsafe_removal_pain_multiplier*I.w_class)//It hurts to rip it out, get surgery you dingus.
 				I.loc = get_turf(src)
```
all ```thing.loc = X``` should be ```thing.forceMove(X)```

RemieRichards reviewed (changes requested):
Wasn't able to review this fully as github lags like a mother.

couple of `loc = X` lines need changing (didn't comment them all, couldn't due to lag), and one weird line that doesn't actually do anything and never has, regardless of your changes to it.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -220,14 +219,9 @@
-				else
-					mech_damage_cycle++
 				var/obj/mecha/M = target
-				M.take_directional_damage(damage_per_tick, "fire", get_dir(src, M), 0, 1, sending_message)
```
Accidental reversion.

phil235 replied:
not accidental, take_directional_damage() no longer exists.

ChangelingRain replied:
The sending_message var and associated not-spamming-the-target-to-death stuff, though?

phil235 replied:
sacrificed for the greater good :snake:

ChangelingRain replied:
Please don't spam the target to death that's no fun

optimumtact replied:
sacrificed for the greater good

phil235 replied:
I found a compromise: when the mech receive damage lower than 5, it only has a probability of receiving the "you're taking damage" message, this should reduce the spam.

ChangelingRain commented on code/modules/paperwork/photocopier.dm:
```diff
@@ -251,8 +254,7 @@
 				user << "<span class='warning'>The [src] smokes, smelling of brimstone!</span>"
-				resistance_flags |= ON_FIRE
-			else
+				resistance_flags |= ON_FIRE //phil235 ????			else
```
Presumably to prevent copying infernal documents

phil235 commented on code/game/gamemodes/clock_cult/clock_items.dm:
```diff
@@ -639,7 +639,7 @@
 	siemens_coefficient = 0
 	permeability_coefficient = 0.05
-	armor = list(melee = 80, bullet = 50, laser = -15, energy = 0, bomb = 35, bio = 0, rad = 0, fire = 20, acid = 20)
+	armor = list(melee = 80, bullet = 50, laser = -15, energy = 5, bomb = 35, bio = 0, rad = 0, fire = 20, acid = 20)
```
my bad

optimumtact commented:
to review this in the cli

    cd /path/to/your/repo
    git remote add phil https://github.com/phil235/-tg-station.git
    git fetch phil
    //assuming you're on the latest master kept up to date with upstream
    git diff phil/damagerino

optimumtact commented:
also apologies I forgot we were in a merge freeze and merged some things

AnturK commented:
Is there really no acid sound ? Welder seems not really fitting.

phil235 commented:
@AnturK I couldn't find a better fit among the existing sounds.

WJohn commented:
@cuboos to the rescue, please!

Cuboos commented:
@WJohn what do you need?

phil235 commented:
@Cuboos A sound for acid melting through objects would be nice.

Cuboos commented:
@phil235 Alright, i've got some school work i need to do, but after that i'll see what i can come up with

Cuboos commented:
https://dl.dropboxusercontent.com/u/9654207/SS13%20previes/acid_sizzle.ogg

How's this?

phil235 commented:
The sound is too long, it should be less than 3 seconds.

phil235 committed:
more work on stuff
including damage overlay when items are damaged. Also shown when the item is clothing worn by a mob.

nullbear commented:
preference for objects to use:

integrity and maxintegrity instead of health and maxhealth.

phil235 committed:
A lot of work again.
Tweaking stuff for the damage of fire.
Adding repair method for clothes.
I tweaked environment_smash var of simple_animals: you can't one shot closets and tables when that var is equal to 1, we use obj_damage to give the mob a decent punch against all objs instead.
Fixes extinguisher puff not washing off acid off your clothes.
Made some objects indestructible for balance purposes and to not ruin certain gamemodes.

phil235 committed:
Even more work.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into damagerino

# Conflicts:
#	code/game/mecha/mecha_defense.dm

nullbear commented:
yas omg phil i love you

phil235 committed:
map fixing

phil235 committed:
more map fixing

phil235 committed:
fixing even more maps

ChangelingRain commented:
I literally can't review this pr at this point, it's too fucking large and touches too many goddamn files for my browser to handle.

optimumtact commented:
maintainers should know how to review from the cli reee

ChangelingRain commented:
i'm GLAD, oranges, that you've VOLUNTEERED to REVIEW this HUGE-ASS PR ALL BY YOURSELF

optimumtact commented:
I gave clear instructions earlier in the thread

phil235 commented on code/modules/VR/vr_sleeper.dm:
```diff
@@ -139,7 +139,7 @@
 			if(UNCONSCIOUS)
 				status = "Unconscious"
-		data["vr_avatar"] = list("name" = vr_human.name, "status" = status, "health" = vr_human.health, "maxhealth" = vr_human.maxHealth)
+		data["vr_avatar"] = list("name" = vr_human.name, "status" = status, "health" = vr_human.health, "max_integrity" = vr_human.maxHealth)
```
tiny mistake here

optimumtact commented:
From my cli review, some things that stood out

```
 /obj/machinery/flasher/Move()
@@ -53,7 +47,7 @@
                else
                        icon_state = "[base_state]1"
        else
-               stat |= NOPOWER
+               stat |= ~NOPOWER
                icon_state = "[base_state]1-p"
```
Was this simply a bug in the original code or did you accidentally invert the flag?
```
 /obj/machinery/shieldwall
...
-       var/obj/machinery/shieldwallgen/gen_secondary
...
+               name = "shield"
```
You also appeared to double tab indent a bunch of type definitions in the shield code(example above)

```
+++ b/code/game/machinery/suit_storage_unit.dm
@@ -1,4 +1,4 @@
-// SUIT STORAGE UNIT /////////////////
+// SUIT STORAGE UNIT /////////////////close_machine(^M
```
stray proc name after the comments (not really a problem)

```
-/obj/item/device/taperecorder/fire_act(exposed_temperature, exposed_volume)
+/obj/item/device/taperecorder/fire_act()^M
        mytape.ruin() //Fires destroy the tape
-       ..()
+       return()^M
```
return() 

```
+/obj/structure/grille/Deconstruct()
+       if(!loc) //if already qdel'd somehow, we do nothing
+               return
```
You should check this with if(qdeleted(src)) instead of relying on the loc being null

```
 /obj/item/device/modular_computer/proc/break_apart()
...
-       qdel(src)
+       qdel()
```
Does this even work?

```
-               loaded_item.forceMove(loc)
+               loaded_item.loc = loc^M
```
You seem to have removed forceMoves in multiple locations in the code, was was your reasoning behind that?


other than minor concerns like this, the overall refactor is solid and reduces a lot of the code, especially in the mob life and damage calculations.

phil235 commented:
@optimumtact 
```
-               stat |= NOPOWER
+               stat |= ~NOPOWER
```
I don't know what you did, but it's the other way around in the code, I removed the rogue ~ in the PR, not the other way around.

optimumtact commented:
woah

Militaires commented:
does this make xeno blood spicy

pls do

phil235 committed:
Some requested changes and fixes.

phil235 commented:
@Militaires Nope.

Militaires commented:
@phil235 tho with the implementation of acid damage and stepping on acid now is the best time to add it or fully flesh it out verbally

phil235 commented:
@Militaires It'd be hard to implement since the acid effect is /obj/effect/acid whereas alien blood is obj/effect/decal/cleanable/bloodstuffhere , I'd have to copypaste all acid code there, or rewrite xenoblood path to be completely different with various consequences.

Also, the PR is ready, I added the changelog, fixed the requested things and typos.

Militaires commented:
i see, nonetheless excellent pr @phil235 
dis gud

now please merge it and cure github of constipation

Militaires commented:
i see, nonetheless excellent pr @phil235 
dis gud

now please merge it and cure github of constipation

although you CAN make xenos spawn the effect/acid when damaged beyond a certain threshold, that too would work

phil235 commented:
>although you CAN make xenos spawn the effect/acid when damaged beyond a certain threshold, that too would work

That's actually a  pretty good idea.

optimumtact commented:
Lets not put any more into this pr so we can get it merged

phil235 commented:
Of course, I'm not putting that idea in THIS pr. god no.

Core0verload commented:
Did this reverted mech directional armor?

phil235 commented:
@coiax No, look at the take_damage() proc, it has a direction as argument, specifically to keep this functionality.
